url2bc31e3489_ramonli_RabbitMQResearch_tgz-pJ8-com_mpos_lottery_te_thirdpartyservice_amqp_protobuf_CancellationMessagePulisherJ8/src/com/mpos/lottery/te/thirdpartyservice/amqp/RabbitMessagePublisher.java

```diff
@@ -0,0 +1,111 @@
+package com.mpos.lottery.te.thirdpartyservice.amqp;
+
+import java.io.IOException;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.apache.commons.pool.ObjectPool;
+
+import com.mpos.lottery.te.config.exception.SystemException;
+import com.rabbitmq.client.Channel;
+import com.rabbitmq.client.Connection;
+import com.rabbitmq.client.MessageProperties;
+import com.rabbitmq.client.ShutdownSignalException;
+
+public class RabbitMessagePublisher implements MessagePublisher {
+    private Log logger = LogFactory.getLog(RabbitMessagePublisher.class);
+    private ObjectPool connectionPool;
+
+    public RabbitMessagePublisher(ObjectPool connPool) {
+        this.connectionPool = connPool;
+    }
+
+    @Override
+    public void publish(final byte[] message, final String exchangeName, final String routingKey)
+            throws IOException {
+        this.exchange(new MessageCallback() {
+
+            @Override
+            public void doExchange(MessageContext context) throws ShutdownSignalException, IOException {
+                Channel channel = context.getChannel();
+                // declare a 'topic' type of exchange
+                channel.exchangeDeclare(exchangeName, "topic");
+                // Content-type "application/octet-stream", deliveryMode 2
+                // (persistent), priority zero
+                channel.basicPublish(exchangeName, routingKey, MessageProperties.PERSISTENT_BASIC, message);
+                if (logger.isDebugEnabled())
+                    logger.debug("Publish message[" + message + "] successfully.");
+            }
+        });
+    }
+
+    /**
+     * This method will manage the connection pool, and re-connect is remote
+     * server is down. The real publishing or consuming is delegated to
+     * <code>MessageCallback</code>.
+     * 
+     * @param callback
+     *            THe <code>MessageCallback</code> which will publish or consume
+     *            message.
+     */
+    protected void exchange(MessageCallback callback) throws IOException {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 11. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.